### PR TITLE
Feat/vector db pgvector

### PR DESCRIPTION
**Pull request by vaayne**
# Description

Support [pgvector](https://github.com/pgvector/pgvector) a new type of vector database.

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Local integration tests

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by bowenliang123** [comment]
Please provide reasonable description for your suggeseted changes, design and approaches.

And this PR seems likely duplicated to #3833, which is in a more maturity in implementation.

**Comment by JohnJyong** [comment]
hi @vaayne , pls add the integration_tests for the pgvector  in the folder 'tests.integration_tests.vdb', thanks

**Comment by vaayne** [comment]
all done @JohnJyong

**Comment by JohnJyong** [comment]
This branch has conflicts that must be resolved @vaayne

**Comment by JohnJyong** [comment]
our db is using the postgresql ,  the pgvector is using the same volumes path  and port with it , it's better to set another folder path and port  for pgvector , thanks @vaayne

**Comment by vaayne** [comment]
All done @JohnJyong

File: api/core/rag/datasource/vdb/pgvector/pgvector.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            values.append(
+                (
+                    doc_id,
+                    doc.page_content,
+                    json.dumps(doc.metadata),
+                    embeddings[i],
+                )
+            )
+        with self._get_cursor() as cur:
+            cur.executemany(
```
Comment (ThreeFish-AI) [inline api/core/rag/datasource/vdb/pgvector/pgvector.py]: Consider using `psycopg2.extras.execute_values()` here?
In bulk insert scenarios, `executemany()` seems to be dozens of times slower than `execute_values()`.

Comment (vaayne) [reply api/core/rag/datasource/vdb/pgvector/pgvector.py]: great suggestion

**Comment by vaayne** [comment]
@JohnJyong Do you have time to review this PR, Thanks

File: api/core/rag/datasource/vdb/pgvector/pgvector.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    def delete_by_ids(self, ids: list[str]) -> None:
+        with self._get_cursor() as cur:
+            cur.execute(f"DELETE FROM {self.table_name} WHERE id IN %s", (tuple(ids),))
+
+    def delete_by_metadata_field(self, key: str, value: str) -> None:
+        with self._get_cursor() as cur:
+            cur.execute(f"DELETE FROM {self.table_name} WHERE meta->>%s = %s", (key, value))
+
+    def search_by_vector(self, query_vector: list[float], **kwargs: Any) -> list[Document]:
```
Comment (JohnJyong) [inline api/core/rag/datasource/vdb/pgvector/pgvector.py:124]: the score is not returned in this method and we should use ' cosine' as the default value of distance_metric.
example code for pgvecto_rs, maybe it is useful for you  @vaayne 
```
        with Session(self._client) as session:
            stmt = (
                select(
                    self._table,
                    self._table.vector.op(self._distance_op, return_type=Float)(
                        query_vector,
                    ).label("distance"),
                )
                .limit(kwargs.get('top_k', 2))
                .order_by("distance")
            )
            res = session.execute(stmt)
            results = [(row[0], row[1]) for row in res]

        # Organize results.
        docs = []
        for record, dis in results:
            metadata = record.meta
            score = 1 - dis
            metadata['score'] = score
            score_threshold = kwargs.get('score_threshold') if kwargs.get('score_threshold') else 0.0
            if score > score_threshold:
                doc = Document(page_content=record.text,
                               metadata=metadata)
                docs.append(doc)
        return docs
```

File: api/core/rag/datasource/vdb/pgvector/pgvector.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        cache_key = f"vector_indexing_{self._collection_name}"
+        lock_name = f"{cache_key}_lock"
+        with redis_client.lock(lock_name, timeout=20):
+            collection_exist_cache_key = f"vector_indexing_{self._collection_name}"
+            if redis_client.get(collection_exist_cache_key):
+                return
+
+            with self._get_cursor() as cur:
+                cur.execute("CREATE EXTENSION IF NOT EXISTS vector")
+                cur.execute(f"DROP TABLE IF EXISTS {self.table_name}")
```
Comment (JohnJyong) [inline api/core/rag/datasource/vdb/pgvector/pgvector.py]:
```
cur.execute(f"DROP TABLE IF EXISTS {self.table_name}")
```
pls delete this logic @vaayne

File: api/core/rag/datasource/vdb/pgvector/pgvector.py (hunk 2 above)
Comment (vaayne) [reply api/core/rag/datasource/vdb/pgvector/pgvector.py:124]: fixed

**Comment by JohnJyong** [comment]
LGTM , thanks for your contribution @vaayne

**Review by JohnJyong — APPROVED** [review]

**Pull request merged**